tests/services/sessions/test_api.py

```diff
@@ -1,61 +1,307 @@
+        return self._req(id, method='PATCH', body=body)
 
+    def modify_type(self, id, type):
+        body = {'type': type}
+        return self._req(id, method='PATCH', body=body)
 
-async def test_create(fetch):
+    def modify_kernel_name(self, id, kernel_name):
+        body = {'kernel': {'name': kernel_name}}
+        return self._req(id, method='PATCH', body=body)
+
+    def modify_kernel_id(self, id, kernel_id):
+        # Also send a dummy name to show that id takes precedence.
+        body = {'kernel': {'id': kernel_id, 'name': 'foo'}}
+        return self._req(id, method='PATCH', body=body)
+
+    async def delete(self, id):
+        return await self._req(id, method='DELETE')
+
+    async def cleanup(self):
+        resp = await self.list()
+        sessions = j(resp)
+        for session in sessions:
+            await self.delete(session['id'])
+        time.sleep(0.1)
+
+
+
+@pytest.fixture
+def session_client(root_dir, fetch):
+    subdir = root_dir.joinpath('foo')
+    subdir.mkdir()
+
+    # Write a notebook to subdir.
+    nb = new_notebook()
+    nb_str = writes(nb, version=4)
+    nbpath = subdir.joinpath('nb1.ipynb')
+    nbpath.write_text(nb_str, encoding='utf-8')
+
+    # Yield a session client
+    client = SessionClient(fetch)
+    yield client
+
+    # Remove subdir
+    shutil.rmtree(str(subdir), ignore_errors=True)
+    
+
+async def test_create(session_client):
     # Make sure no sessions exist.
-    r = await fetch(
-        'api', 'sessions',
-        method='GET'
-    )
-    sessions = json.loads(r.body.decode())
+    resp = await session_client.list()
+    sessions = j(resp)
     assert len(sessions) == 0
 
     # Create a session.
-    model = get_session_model('foo/nb1.ipynb')
-    r = await fetch(
-        'api', 'sessions',
-        method='POST',
-        body=json.dumps(model)
-    )
-    assert r.code == 201
-    new_session = json.loads(r.body.decode())
+    resp = await session_client.create('foo/nb1.ipynb')
+    assert resp.code == 201
+    new_session = j(resp)
     assert 'id' in new_session
     assert new_session['path'] == 'foo/nb1.ipynb'
     assert new_session['type'] == 'notebook'
-    assert r.headers['Location'] == '/api/sessions/' + new_session['id']
+    assert resp.headers['Location'] == '/api/sessions/' + new_session['id']
 
     # Check that the new session appears in list.
-    r = await fetch(
-        'api', 'sessions',
-        method='GET'
-    )
-    sessions = json.loads(r.body.decode())
+    resp = await session_client.list()
+    sessions = j(resp)
     assert sessions == [new_session]
 
     # Retrieve that session.
     sid = new_session['id']
-    r = await fetch(
-        'api', 'sessions', sid,
-        method='GET'
-    )
-    got = json.loads(r.body.decode())
+    resp = await session_client.get(sid)
+    got = j(resp)
     assert got == new_session
 
+    # Need to find a better solution to this.
+    await session_client.cleanup()
```

Comment:
Can we add these cleanup calls to the fixture, after the yield and prior to the dir removal?

Reply:
Yeah, I had tried to get this to work, but async-ness of this call makes it challenging. I couldn’t find a way to both yield in the session client fixture and use this asynchronous cleanup method. Do you how to do this?

Reply:
Yeah, I see what you mean.  I just tried a few things to no avail.  There are ways to create async fixtures but I'm not having much luck.  Let's revisit since this is more of a test impl issue.

Reply:
I'm currently having success with this (using pytest.yield_fixture and run_sync).  File attached below ...
```python

def run_sync(coro_method):
    return asyncio.get_event_loop().run_until_complete(coro_method)


@pytest.yield_fixture
def session_client(root_dir, fetch):
    subdir = root_dir.joinpath('foo')
    subdir.mkdir()

    # Write a notebook to subdir.
    nb = new_notebook()
    nb_str = writes(nb, version=4)
    nbpath = subdir.joinpath('nb1.ipynb')
    nbpath.write_text(nb_str, encoding='utf-8')

    # Yield a session client
    client = SessionClient(fetch)
    yield client

    # Invoke async method on separate loop
    run_sync(client.cleanup())

    # Remove subdir
    shutil.rmtree(str(subdir), ignore_errors=True)
```
[test_api.py.txt](https://github.com/jupyter/jupyter_server/files/3935489/test_api.py.txt)

Reply:
Warning - I'm not convinced this will work on 3.5 and I'd prefer to understand things better (like why does yield_fixture change things).